src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -772,10 +774,78 @@ impl ToStream for LogicalJoin {
                 Ok(StreamHashJoin::new(logical_join, predicate).into())
             }
         } else {
-            Err(RwError::from(ErrorCode::NotImplemented(
+            let nested_loop_join_error = RwError::from(ErrorCode::NotImplemented(
                 "stream nested-loop join".to_string(),
                 None.into(),
-            )))
+            ));
+            // If there is exactly one predicate, it is a comparison (<, <=, >, >=), and the
+            // join is a `Inner` join, we can convert the scalar subquery into a
+            // `StreamDynamicFilter`
+
+            // Check if `Inner` subquery (no `IN` or `EXISTS` keywords)
+            if self.join_type != JoinType::Inner {
+                return Err(nested_loop_join_error);
+            }
+
+            // Check if right side is a scalar (for now, check if it is a simple agg)
+            let maybe_simple_agg = if let Some(proj) = self.right().as_logical_project() {
+                proj.input()
+            } else {
+                self.right()
+            };
+
+            if let Some(agg) = maybe_simple_agg.as_logical_agg() && agg.group_keys().is_empty() {
+                /* do nothing */
+            } else {
+                return Err(nested_loop_join_error);
+            }
+
+            // Check if the join condition is a correlated comparison
+            let conj = &predicate.other_cond().conjunctions;
+
+            let left_ref_index = if let Some(expr) = conj.first() && conj.len() == 1
+            {
+                if let Some((left_ref, _, right_ref)) = expr.as_comparison_cond()
+                    && left_ref.index < self.left().schema().len()
+                    && right_ref.index >= self.left().schema().len()
+                {
+                    left_ref.index
+                } else {
+                    return Err(nested_loop_join_error);
+                }
+            } else {
+                return Err(nested_loop_join_error);
+            };
+
+            let left = self
+                .left()
+                .to_stream_with_dist_required(&RequiredDist::shard_by_key(
+                    self.left().schema().len(),
+                    &[left_ref_index],
+                ))?;
+
+            assert!(*self.right().distribution() == Distribution::Single);
+
+            let right = self
+                .right()
+                .to_stream_with_dist_required(&RequiredDist::PhysicalDist(
+                    Distribution::Broadcast,
+                ))?;
+
+            let plan = StreamDynamicFilter::new(predicate.other_cond().clone(), left, right).into();
+
+            if self.output_indices != (0..self.internal_column_num()).collect::<Vec<_>>() {
```

Comment:
Actually, we should probably use the rewritten logical join and output indices

Reply:
actually, they are identical, so this is unnecessary. simplified some existing code